### PR TITLE
switch to multiplatorm

### PR DESCRIPTION
**Pull request by ethanmdavidson**
This PR is aimed at making this library work on multiple platforms. I've switched to the multiplatform plugin and rearranged the code accordingly, and all the tests are passing.

I'm not sure how difficult it will be to get the publishing to work. [The documentation](https://kotlinlang.org/docs/mpp-publish-lib.html) is a little sparse. I also found [this example repo](https://github.com/KaterinaPetrova/mpp-sample-lib/blob/main/convention-plugins/src/main/kotlin/convention.publication.gradle.kts). I've heard that the gradle plugin takes care of everything for you, but I've never done it before. Please let me know how I can help with this.

### CODE REVIEW

**Comment by ethanmdavidson** [comment]
It's my understanding that MavenCentral is capable of hosting the multiplatform artifacts (again, I've never actually done it). If you run `./gradlew build` you'll see several jar files in `build/libs/`, which I think include the stuff that is needed for all the platforms (including js).

**Comment by ethanmdavidson** [comment]
After making a small change to the maven publication config, I'm able to run `./gradlew publishJsPublicationToMavenLocal` and then use this library in a kotlin/js project (just using `publishToMavenLocal` didn't work for some reason). I think you might need to use the `publishAllPublicationsToMavenRepository` target.

**Comment by notKamui** [comment]
Again, thanks a lot for your contribution. Gonna check that later today and merge the PR

**Comment by notKamui** [comment]
I added compilation targets for linux and mingw ; overall it seems fine (i managed to publish to my snapshot repository)

However, I do receive a message, which I expected :


In the documentation you linked, they talk about uploading duplicates and how to prevent them. If we don't do that, I won't be able to publish it the main repository as "smoothly" as possible.

**Comment by ethanmdavidson** [comment]
I think the duplicates thing only applies if you need to build on multiple machines, e.g. if it targets jvm, macos, and mingw it will need to be built on a mac machine (for the mac target) and a windows machine (for the mingw target), but because jvm and js can be built on both machines you need to make sure they only get built+published once, so you restrict them to only build on one host. In other words, I don't think it applies to this project, unless the linuxX64 and mingwX64 targets need to be built on different hosts?

However, I think maybe the duplicate warning you got might be indicating that we're using the wrong gradle task. I think I may have been mistaken about `publishAllPublicationsToMavenRepository`, since in the snapshot repo there are multiple publications (`keval-0.7.5-20211217.131800-1`, `keval-0.7.5-20211217.131800-2`, `keval-0.7.5-20211217.131800-3`, `keval-0.7.5-20211217.131800-4`, `keval-0.7.5-20211217.131800-5`). When I try to include the 0.7.5-SNAPSHOT version as a js dependency, it seems to only use the last one, which doesn't include the JS platform. I'm guessing these 5 publications are 1 per platform specified (jvm, js-browser, js-node, linux, mingw). Looking at the module info for the last publication (https://oss.sonatype.org/content/repositories/snapshots/com/notkamui/libs/keval/0.7.5-SNAPSHOT/keval-0.7.5-20211217.131800-5.module) does seem to indicate that it only includes the mingw package.

Could you try publishing the snapshot again, but just using the `publish` target?

**Comment by notKamui** [comment]
> Could you try publishing the snapshot again, but just using the `./gradlew publish` task?

Hmm, I've just done so, and the same problem still persists, unfortunately. I still get the warnings, as well as the fact that each platform gets its own publication.

At least we're learning by trial and errors.

https://kotlinlang.org/docs/mpp-publish-lib.html#structure-of-publications

Here they talk about a specificity of Maven Central (which I honestly don't quite understand), as well as a supposed root artifact.

Said folder does appear in the build path, with a module, but strangely no jar for it is built.

File: build.gradle.kts [hunk 1]
```diff
@@ -56,15 +73,16 @@ tasks {
...
     "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
 
 publishing {
     publications {
-        create<MavenPublication>("mavenJava") {
-            from(components["java"])
+        withType<MavenPublication> {
             groupId = project.group.toString()
             artifactId = project.name.toLowerCase()
             version = project.version.toString()
```
Comment (martinbonnin) [inline build.gradle.kts:88]: The Kotlin multiplatform plugin will set these for you as there can be `Keval-jmv`, `Keval-js`, etc... If you remove these, the warning should go away

Comment (notKamui) [reply build.gradle.kts:88]: Great, gonna check that later !

**Comment by notKamui** [comment]
Ok so now after the suggestion made by @martinbonnin we have this
image

and the module in Keval/ does contain every targets ! Now I wish I could rename the artifactId manually but I guess I'm just gonna have to reset the project name to `keval` instead of `Keval`. This is fine to me, but do you know if there's a way to make it
use the lowercase of the project name ?

**Comment by ethanmdavidson** [comment]
I can confirm that I was able to include 0.7.5.1-SNAPSHOT in a KMP project and have it work for the js target!

I think my last commit should force the artifactId to lowercase, while leaving the project uppercase.

**Comment by notKamui** [comment]
Indeed, I didn't think of using the already set artifactId ! That works like a charm ! Ready to merge

Thanks again :)

**Pull request merged**